crates/daphne/src/pine/flp.rs

```diff
@@ -174,27 +174,54 @@ impl<F: FftFriendlyFieldElement> PineType<F> {
         gradient: &[F],
         wr_joint_rand_seed: &Seed<16>,
     ) -> [F; NUM_WR_TESTS] {
+        debug_assert_eq!(gradient.len(), self.cfg.dimension);
         let mut xof = XofTurboShake128::seed_stream(
             wr_joint_rand_seed,
             &self.cfg.dst(USAGE_WR_JOINT_RAND),
             &[],
         );
-        let mut buf = vec![0_u8; chunk_count(4, NUM_WR_TESTS * self.cfg.dimension)];
-        xof.fill(&mut buf[..]);
+        let rand_len_per_test = chunk_count(4, self.cfg.dimension);
+        let mut rand = vec![0_u8; rand_len_per_test * NUM_WR_TESTS];
+        xof.fill(&mut rand[..]);
 
         let mut wr_test_results = [F::zero(); NUM_WR_TESTS];
-        let mut i = 0;
-        for wr_test_result in &mut wr_test_results {
-            for x in gradient {
-                // TODO spec: Consider reversing the order in which we read the byte. We can save a
-                // little computation if we can get rid of the subtraction below.
-                let rand_bits = (buf[i >> 3] >> (6 - (i & 7))) & 0b11;
-                match rand_bits {
-                    0b00 => *wr_test_result -= *x,
-                    0b11 => *wr_test_result += *x,
+        for (wr_test_result, rand_per_test) in wr_test_results
+            .iter_mut()
+            .zip(rand.chunks(rand_len_per_test))
+        {
+            for (gradient_chunk, r) in gradient.chunks(4).zip(rand_per_test) {
+                match r & 3 /* 0b11 */ {
+                    0 => *wr_test_result -= gradient_chunk[0],
+                    3 => *wr_test_result += gradient_chunk[0],
+                    _ => (),
+                };
+
+                if gradient_chunk.len() < 2 {
+                    continue;
+                }
```

Comment:
if- sentences is what reduce preformance on programs, so what if you use [exact_chunks](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks_exact), so the loop has no branches, and only process the remainder bytes one at a time.

Reply:
I tacked on a commit with this change. Interestingly there is no change in performance. I suspect this is because the CPU is predicting that the `if` branches aren't taken?

Reply:
Likely, however with this change we've removed the branch prediction logic.